Azure : filter disks with ToBeDetached flag

kkmsft opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change

 /kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Filter disks with ToBeDetached flag

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

andyzhangx commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_vmss.go:
```diff
@@ -42,11 +57,13 @@ func (ss *scaleSet) AttachDisk(isManagedDisk bool, diskName, diskURI string, nod
 	if vm.StorageProfile != nil && vm.StorageProfile.DataDisks != nil {
-		disks = make([]compute.DataDisk, len(*vm.StorageProfile.DataDisks))
-		copy(disks, *vm.StorageProfile.DataDisks)
+		unfilteredDisks = make([]compute.DataDisk, len(*vm.StorageProfile.DataDisks))
```
what about write like following, then you don't need to copy out to a `unfilteredDisks`
```
filteredDisks = make([]compute.DataDisk, len(*vm.StorageProfile.DataDisks))
filterDetatchingDisks(*vm.StorageProfile.DataDisks, filteredDisks)
```

```
func filterDetatchingDisks( ...) {
  for _, disk := range dataDisks {
      if !*disk.ToBeDetached {
         append disk to filteredDisks
   }
  }
}
```

andyzhangx commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_vmss.go:
```diff
@@ -28,6 +28,21 @@ import (
 	"k8s.io/klog"
 )
 
+func filterDetatchDisks(oldDisks []compute.DataDisk) (disks []compute.DataDisk) {
```
`filterDetatchingDisks` and also write a UT

andyzhangx commented:
/ok-to-test

kkmsft commented:
/retest

andyzhangx commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils.go:
```diff
@@ -0,0 +1,26 @@
+		if disk.ToBeDetached != nil {
+			if *disk.ToBeDetached {
+				if disk.Name != nil {
+					klog.V(2).Infof("Filtering disk: %s with ToBeDetached flag set.", *disk.Name)
```
klog.V(12).Infof

kkmsft commented:
/retest

andyzhangx commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_utils_test.go:
```diff
@@ -0,0 +1,32 @@
+
+func TestFilteredDetatchingDisks(t *testing.T) {
+
+	disks := []compute.DataDisk{
```
pls add more tests, e.g. empty data disk list

andyzhangx reviewed (commented):
for UT, pls refer to TestStandardDetachDisk.

kkmsft commented:
/retest

kkmsft commented:
/retest

kkmsft commented:
/retest

andyzhangx commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go:
```diff
@@ -289,3 +289,23 @@ func (c *controllerCommon) DisksAreAttached(diskNames []string, nodeName types.N
+	for _, disk := range unfilteredDisks {
+		if disk.ToBeDetached != nil {
+			if *disk.ToBeDetached {
+				if disk.Name != nil {
```
since disk is filtered, log level should be V(2), need to call out

andyzhangx replied:
Looks like I suggested V(12) in the above comment, maybe V(2) is more appropriate since attach/detach does not always happen, we may need to call out, sorry for the confusion.

andyzhangx commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_common.go:
```diff
@@ -289,3 +289,23 @@ func (c *controllerCommon) DisksAreAttached(diskNames []string, nodeName types.N
+func filterDetachingDisks(unfilteredDisks []compute.DataDisk) []compute.DataDisk {
+	filteredDisks := []compute.DataDisk{}
+	for _, disk := range unfilteredDisks {
+		if disk.ToBeDetached != nil {
```
```
if disk.ToBeDetached != nil  && *disk.ToBeDetached {
     if disk.Name != nil {
         klog.V(2).Infof("Filtering disk: %s with ToBeDetached flag set.", *disk.Name)
     }
} else {
    filteredDisks = append(filteredDisks, disk)
}
```

kkmsft committed:
Azure: Filter disks with ToBeDetached flag from attach/detach

After the detach calls are made, any further updates with the disks in
the process of getting detached will complain with errors such as:

"
Cannot attach data disk '<disk name>' to VM '<vm name>' because the
disk's ToBeDetached flag is set to true. Please either remove the
disk from APIModel or set the ToBeDetached flag to false and then
try again
"

This PR will ensure that the disks with the ToBeDetached flag set are
filtered out before the update to the cloud is made.

kkmsft committed:
Azure: Filter disks with ToBeDetached flag from attach/detach- UT

andyzhangx commented:
/test pull-kubernetes-e2e-aks-engine-azure

andyzhangx commented:
hi @chewong could you take a look at the aks-engine-azure test error? Thanks.
```
error building hyperkube error during make -C /go/src/k8s.io/kubernetes WHAT=cmd/hyperkube: exit status 2
```

chewong commented:
> hi @chewong could you take a look at the aks-engine-azure test error? Thanks.
> 
> ```
> error building hyperkube error during make -C /go/src/k8s.io/kubernetes WHAT=cmd/hyperkube: exit status 2
> ```

Jobs that involve building hyperkube on k8s master branch / PRs won't work because of the incompatibility between the new hyperkube image and aks-engine. https://github.com/kubernetes/test-infra/pull/15190 and https://github.com/Azure/aks-engine/pull/2298 will fix the issue.

andyzhangx commented:
/priority important-soon
/sig cloud-provider
/area provider/azure

feiskyer commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_standard.go:
```diff
@@ -42,8 +42,7 @@ func (as *availabilitySet) AttachDisk(isManagedDisk bool, diskName, diskURI stri
 
-	disks := make([]compute.DataDisk, len(*vm.StorageProfile.DataDisks))
-	copy(disks, *vm.StorageProfile.DataDisks)
+	disks := filterDetachingDisks(*vm.StorageProfile.DataDisks)
```
if the disk is under detaching, then attach would still error, right? could we add a check here and fail earlier?

andyzhangx replied:
I think this PR is on the right way, when there is disk under detaching(actually that issue is already fixed by #85158, ), this RP would remove those detaching disks and go on attaching another disk, this works according to vmss team's spec.

feiskyer commented on staging/src/k8s.io/legacy-cloud-providers/azure/azure_controller_standard.go:
```diff
@@ -117,8 +116,7 @@ func (as *availabilitySet) DetachDisk(diskName, diskURI string, nodeName types.N
 
-	disks := make([]compute.DataDisk, len(*vm.StorageProfile.DataDisks))
-	copy(disks, *vm.StorageProfile.DataDisks)
+	disks := filterDetachingDisks(*vm.StorageProfile.DataDisks)
```
same here, check whether the disk is under detaching and fail earlier

andyzhangx replied:
no need to check

feiskyer commented:
/hold

feiskyer commented:
/hold cancel

andyzhangx reviewed (approved):
/lgtm
@feiskyer any other comments?

andyzhangx reviewed (approved):
/approve
/hold cancel

andyzhangx commented:
/kind bug

feiskyer commented:
looks good, thanks for fix.
